### DISCUSSION TITLE
text-border (based on text-shadow)

### DISCUSSION THREAD

**Discussion by silveltman** [tailwindlabs/tailwindcss#8598] [Ideas]
Something that is often used and not ussported by tailwind: text-borders! Or actually text-shadows without blur behind the scene. 
Like having a 1px (or 2 or whatever) shadow around text that functions as a 1px border. 

I think this is something that's used very often in designs and can be added without a lot of investigation work, because it can use the same rules/guidelines as for normal borders Just using 'text-border' instead of 'border'.

**Comment by rohmann** [comment]
Hi @silveltman,

Cool idea! Maybe it will be a feature at some point, but if you want to try it sooner I put together a plugin that will do this. You can start with `text-border` and add any Tailwind color like `text-border-red-900`. Optionally, `text-border-size-2` could be added to adjust the size. It uses the color and scale values for borders out of your theme.

The plugin code is below, and you can also play with it here: https://play.tailwindcss.com/PN21K8yytB?file=config


```js
const plugin = require('tailwindcss/plugin')
const flattenColorPalette = require('tailwindcss/src/util/flattenColorPalette')
const toColorValue = require('tailwindcss/src/util/toColorValue')

module.exports = {
  plugins: [
    plugin(function ({ matchUtilities, e, config, theme }) {
      const textBorderSize = `--tw${config('prefix')}-text-border-size`

      matchUtilities(
        {
          'text-border': (value) => ({
            'text-shadow': `0 0 var(${textBorderSize},1px) ${toColorValue(value)}`,
          }),
        },
        {
          values: (({ DEFAULT: _, ...colors }) => colors)(flattenColorPalette(theme('borderColor'))),
          type: 'color',
        }
      )

      matchUtilities(
        {
          'text-border-size': (value) => ({
            [textBorderSize]: value
          }),
        },
        { values: theme('borderWidth') }
      )
    }),
  ],
}

**Comment by silveltman** [comment]
Wow thats fast, awesome!

In the meantime I've been working on some custom utilities myself which got me to some things that can be applied in this plugin.
I'm not familiar with writing plugins myself, but I have some things that I think can improve the plugin you made. Maybe you can apply it in the plugin?

Some of the custom utilities I made:
```
  .text-border-white {
    color: theme('colors.black.DEFAULT');
    text-shadow: 1px 1px theme('colors.white.DEFAULT'), -1px -1px theme('colors.white.DEFAULT');
  }

  @supports(-webkit-text-stroke: 1px theme('colors.white.DEFAULT')) {
    .text-border-white{
      text-shadow: none;
      color: transparent;
      -webkit-text-stroke: 1px theme('colors.white.DEFAULT');
    }
  }
```

I think the way text shadow is defined here is definitly better, since it does not have a blur and therefore looks more like a border.  In the places where it says '1px' it should still be able to accept the text-border-size (maybe remove the word 'size' here? so it's more like a regular border)

That line would probably look something like this?
`var(${textBorderSize},1px) var(${textBorderSize},1px) 0 ${toColorValue(value)}, var(${textBorderSize},-1px) var(${textBorderSize},-1px) 0 ${toColorValue(value)}`


Another thing I like about my approach is using @supports, because the -webkit-text-stroke makes a border that just looks way better when supported. 

Also, this makes it possible to use text-transparent, which can't be used with text-shadow because this shadow fills the whole text. Tho this might be a bit dangerous, since people are probably developing in a webkit browser... On the other hand, caniuse seems say the support is quiet big. Any thoughts on this part?

**Comment by silveltman** [comment]
This way of applying the text-shadow is probably even better since it has borders on all 4 sides:

```
.myclass {
  text-shadow: 
    1px 1px 0 black, 
    -1px -1px 0 black,
    1px -1px 0 black,
    -1px 1px black;
}
```

**Comment by crswll** [comment]
Depending on the font `text-stroke` can get really strange. Can do something really excessive as well: https://play.tailwindcss.com/jR5fPlbpAo?file=config

**Comment by rohmann** [comment]
@silveltman maybe try your hand at a custom plugin based on some ideas here. Your custom utilities are pretty close and using a plugin will help you get the colors populated automatically.

> That line would probably look something like this?
`var(${textBorderSize},1px) var(${textBorderSize},1px) 0 ${toColorValue(value)}, var(${textBorderSize},-1px) var(${textBorderSize},-1px) 0 ${toColorValue(value)}`

Yes! But instead of -1px you could do `calc(var(${textBorderSize}) * -1)` to use the same size but in the opposite direction (see @crswll's example) 

@crswll That setup looks a whole lot better applied as multiple borders using x/y offsets. (updated my example to properly import the util functions like yours)

**Reply by crswll** [reply]
I'm not positive that's 100% all right to use flattenColorPalette but I sneak it in for experiment stuff like this haha.

**Reply by rohmann** [reply]
well hey, more readable for sure than copy/paste. 😂

**Comment by silveltman** [comment]
I think I've made something pretty solid. It behaves just like normal border would, with the classes 'text-border-[color]' and 'text-border-[width]'

Would love to get some feedback! If everything seems ok I will publish it to npm.

Tailwind play: https://play.tailwindcss.com/zItN8mIHT3?file=config

```
const plugin = require('tailwindcss/plugin')
const flattenColorPalette = require('tailwindcss/src/util/flattenColorPalette')

const textBorder = plugin(
  function ({ matchUtilities, theme, config, e }) {
    const textBorderWidth = `--tw${config('prefix')}-text-border-width`
    const pos = `var(${textBorderWidth},1px)`
    const neg = `calc(var(${textBorderWidth}, 1px) * -1)`

    matchUtilities(
      {
        'text-border': (value) => ({
          'text-shadow': `
            ${pos} ${pos} 0 ${value},
            ${neg} ${neg} 0 ${value},
            ${pos} ${neg} 0 ${value},
            ${neg} ${pos} ${value};
          `,
        }),
      },
      {
        values: flattenColorPalette(theme('colors')),
        type: ['color', 'any'],
      }
    )

    matchUtilities(
      {
        'text-border': (value) => ({
          [textBorderWidth]: value
        }), 
      },
      { values: theme('borderWidth') }
    ) 
  },
)

module.exports = {
  plugins: [
    textBorder
  ],
}
```

**Reply by rohmann** [reply]
Well done! Looks awesome. 👍

**Comment by silveltman** [comment]
I've tried to publish the plugin, but I come a cross some problems. Creating plugins, or npm packages even, is pretty new to me.

It's here on npm: https://www.npmjs.com/package/tailwindcss-text-border

my error:
```
C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\src\util\flattenColorPalette.js:13
export default flattenColorPalette
^^^^^^

SyntaxError: Unexpected token 'export'
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1033:15)
    at Module._compile (node:internal/modules/cjs/loader:1069:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss-text-border\index.js:2:29)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
ERROR: "postcss:watch" exited with 1.
```

Can one of you guys help me out? When I get this to work I will publish a refactored version of this plugin for working with regular text-shadows, as mentioned in  https://github.com/tailwindlabs/tailwindcss/issues/162

**Reply by rohmann** [reply]
Not sure exactly what's up there, but I'd try this first. Instead of importing from `src/util` try `lib/util`. That should allow importing the built modules which use `exports.default` instead of export default`.

So that would be:
```
const flattenColorPalette = require('tailwindcss/src/util/flattenColorPalette')
```

instead of:

```
const flattenColorPalette = require('tailwindcss/lib/util/flattenColorPalette')
```


Also, you might want to define tailwind as a peer dependency. It may help to look over how some of the official plugins ([line-clamp for example(https://github.com/tailwindlabs/tailwindcss-line-clamp)] that are published on npm are setup to get an idea of how that all works.

**Comment by silveltman** [comment]
New error this time :(

```
^ZTypeError: flattenColorPalette is not a function
    at C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss-text-border\index.js:21:15
    at registerPlugins (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\lib\setupContextUtils.js:660:61)
    at createContext (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\lib\setupContextUtils.js:854:5)
    at Object.getContext (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\lib\setupContextUtils.js:910:19)
    at C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\lib\setupTrackingContext.js:40:53
    at C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\processTailwindFeatures.js:38:11
    at plugins (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\tailwindcss\lib\index.js:30:63)
    at LazyResult.runOnRoot (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\postcss\lib\lazy-result.js:339:16)
    at LazyResult.runAsync (C:\Users\silve\Documents\Web dev\11ty\webtunus\node_modules\postcss\lib\lazy-result.js:393:26)
    at async Promise.all (index 0)
^Z
```

**Reply by rohmann** [reply]
Hmm, ok. Give this a try...

```
const flattenColorPalette = require('tailwindcss/lib/util/flattenColorPalette').default
```

**Comment by silveltman** [comment]
That did the trick, thanks!

However, it seems a bit wonky to me this way. Like it could break at any time haha

**Reply by rohmann** [reply]
I don't think that one's part of the public API so it could break with a future tailwind update.

**Reply by crswll** [reply]
Yeah, that's definitely a concern and why I wrote "I'm not positive that's 100% all right to use flattenColorPalette but I sneak it in for experiment stuff like this haha."

I could have sworn I saw it in an official plugin but just looked around and didn't see anything. It's probably better to use what @rohmann originally had.

**Comment by schimin** [comment]
Not working well to me.
When i create a plugin file and paste the code, and import via tailwind.config.ts not work.
Everything in code "tailwind intelissense" stop working.
